backend/app/lib/oai/aspace_oai_repository.rb

```diff
@@ -62,7 +62,22 @@ def sets
       OAI::Set.new(:name => level, :spec => level)
     }
 
-    config_sets + level_sets
+    config_sets + level_sets.select {|s| set_enabled?(s) }
+  end
+
+  # returns true if set is enabled in at least one repository
+  def set_enabled?(set)
+    sets_in_repos = Repository.exclude(:publish => 0)
+                              .exclude(:oai_is_disabled => 1)
+                              .select(:oai_sets_available)
+                              .map {|r| JSON::parse(r[:oai_sets_available]) rescue [] }
+
+    # if oai_sets_enabled array is blank, then all sets are enabled for that repo
```

Comment:
Did you mean oai_sets_available array is blank and not oai_sets_enabled array is blank in the comment? Not sure I understand the comment regardless - can you elaborate more?

Reply:
Oops, there's a typo in the comment -- it's oai_sets_available that we're checking for blankness. 

If oai_sets_available is blank, that means that all sets are enabled for that repository. If only some sets are enabled for a repository, then the set ids are encoded in JSON in that field. I did it that way as an optimization, since in most cases I'd expect a repo to have all sets enabled, and it's easy to tell when this is the case and fall out any set logic in this case.

I'll fix the comment and update it to make it clearer.